lib/yocaml/runtime.ml

```diff
@@ -1,84 +1,103 @@
+
+  let create_predicate path pred = function
+    | `Files ->
+      fun x ->
+        let p = x |> into path in
+        let* file_exists = R.file_exists p in
+        let* is_directory = R.is_directory p in
+        if file_exists && (not is_directory) && pred x
+        then R.return (Some p)
+        else R.return None
+    | `Directories ->
+      fun x ->
+        let p = x |> into path in
+        let* file_exists = R.file_exists p in
+        let* is_directory = R.is_directory p in
+        if file_exists && (not is_directory) && pred x
+        then R.return (Some p)
+        else R.return None
+    | `Both ->
+      fun x ->
+        let p = x |> into path in
+        if pred x then R.return (Some p) else R.return None
+  ;;
 
-let execute (module R : RUNTIME) program =
-  Effect.run
-    { handler =
-        (fun resume effect ->
-          let f : type b. (b -> 'a) -> b Effect.f -> 'a =
-           fun resume -> function
-            | Effect.Get_modification_time path ->
-              resume $ R.get_modification_time path
-            | Effect.Target_modification_time path ->
-              resume $ R.target_modification_time path
-            | Effect.File_exists path -> resume $ R.file_exists path
-            | Effect.Target_exists path -> resume $ R.target_exists path
-            | Effect.Read_file path -> resume $ R.read_file path
-            | Effect.Content_changes (path, content) ->
-              let result =
-                R.content_changes path content
-                |> Try.Functor.map (function
-                       | true -> Either.left content
-                       | false -> Either.right ())
-              in
-              resume result
-            | Effect.Write_file (path, content) ->
-              let () = R.create_dir $ Filename.dirname path in
-              resume $ R.write_file path content
-            | Effect.Read_dir (path, kind, predicate) ->
-              let children = R.read_dir path in
-              let full_predicate =
-                create_predicate (module R) path predicate kind
-              in
-              resume $ List.filter_map full_predicate children
-            | Effect.Log (level, message) ->
-              let () = R.log level message in
-              resume ()
-            | Effect.Throw error ->
-              let () =
-                R.log Log.Alert (Lexicon.crap_there_is_an_error error)
-              in
-              Error.raise' error
-            | Effect.Raise exn ->
-              let () =
-                R.log Log.Alert (Lexicon.crap_there_is_an_exception exn)
-              in
-              raise exn
-          in
-          f resume effect)
-    }
-    program
-;;
+  let filter_map f l =
+    let rec go acc = function
+      | [] -> R.return (List.rev acc)
+      | hd :: tl ->
+        let* hd = f hd in
+        (match hd with
+        | Some v -> (go [@tailcall]) (v :: acc) tl
+        | None -> (go [@tailcall]) acc tl)
+    in
+    go [] l
+  ;;
```

Comment:
Since [`filter_map`](https://ocaml.org/api/List.html#VALfilter_map) is available since `4.08.0` why not using it directly?

Reply:
This `filter_map` uses the `R` monad as `Lwt_list.filter_map` provides. May be a play with `preface` should let us to provide such function.

Reply:
There is a trivial implementation based on Traverse but I think it would add more noise than the proposed implementation!